### PR TITLE
Better hint

### PR DESCRIPTION
**Pull request by abdtyx**
### Main feature
I refactored the resolveSymbols function, now it's easier for students to understand what they need to do with errors.
### Other features
By the way, I also fixed some trivial issue as well:
* Add *.out in gitignore
* The `retProc` in autograder is not used, so I removed it.
* Add a Makefile so that students can run `make` to compile and test their code.
* The `shlen` in util.cc is not used, removed.

### CODE REVIEW

File: src/resolve.cc [hunk 1]
```diff
@@ -1,20 +1,50 @@
 #include "resolve.h"
 
 #include <iostream>
+#include <signal.h>
+#include <set>
+#include <vector>
```
Comment (Qcloud1223) [inline src/resolve.cc]: These headers are already included in `util.h`, and `resolve.h` includes it.

File: src/resolve.cc [hunk 2]
```diff
@@ -1,20 +1,50 @@
...
+
+using std::set, std::string, std::cout, std::endl, std::vector, std::cerr;
```
Comment (Qcloud1223) [inline src/resolve.cc]: I personally don't like it. If our user like it, they can do it themselves.

Comment (abdtyx) [reply src/resolve.cc]: Oops, I forgot to delete them. I used them in my own impl, will fix soon

File: src/resolve.cc [hunk 3]
```diff
@@ -1,20 +1,50 @@
...
+        cerr << "multiple definition for symbol " << ret.sym << endl;
+        abort();
+        break;
+    case errsym::NODEF:
+        cerr << "undefined reference for symbol " << ret.sym << endl;
+        abort();
+        break;
+    default:
+        break;
+    }
```
Comment (Qcloud1223) [inline src/resolve.cc:20]: I don't like passing things around, packing symbol names in another structure and passing it out only for printing. Keep the skeleton code short and clean.

**Review by Qcloud1223 — CHANGES_REQUESTED** [review]
I feel changes in `resolve.cc` is complexing things but not making significant convenience, Other changes are good.

Convince me if you think they are necessary :)

**Comment by abdtyx** [comment]
Just submitted another design, any feedback?
This time the students only need to return the right value and set the `errSymName` which is a private variable within the `resolve.cc`.

**Comment by abdtyx** [comment]
Add two commits

**Review by Qcloud1223 — APPROVED** [review]
LGTM

**Pull request merged**